app/src/main/java/org/tomahawk/tomahawk_android/mediaplayers/AndroidMediaPlayer.java

```diff
@@ -0,0 +1,173 @@
+
+import android.media.AudioManager;
+import android.media.MediaPlayer;
+import android.util.Log;
+
+import org.tomahawk.libtomahawk.resolver.PipeLine;
+import org.tomahawk.libtomahawk.resolver.Query;
+import org.tomahawk.libtomahawk.resolver.Result;
+import org.tomahawk.libtomahawk.resolver.ScriptResolver;
+import org.tomahawk.libtomahawk.utils.NetworkUtils;
+import org.tomahawk.tomahawk_android.utils.ThreadManager;
+import org.tomahawk.tomahawk_android.utils.TomahawkRunnable;
+
+import java.io.IOException;
+import java.util.concurrent.ConcurrentHashMap;
+
+import de.greenrobot.event.EventBus;
+
+
+/**
+ * Created by Anton Romanov on 6/7/16.
+ */
+public class AndroidMediaPlayer implements TomahawkMediaPlayer {
+
+    private static final String TAG = AndroidMediaPlayer.class.getSimpleName();
+
+    private static MediaPlayer sMediaPlayer = null;
+
+    private Query mPreparedQuery;
+
+    private Query mPreparingQuery;
+
+    private final ConcurrentHashMap<Result, String> mTranslatedUrls
+            = new ConcurrentHashMap<>();
+
+    private TomahawkMediaPlayerCallback mMediaPlayerCallback;
+
+    public AndroidMediaPlayer() {
+        EventBus.getDefault().register(this);
+    }
+    @Override
+    public void play() {
+        if (sMediaPlayer != null)
+            sMediaPlayer.start();
+    }
+
+    @Override
+    public void pause() {
+        if (sMediaPlayer != null)
+            sMediaPlayer.pause();
+    }
+
+    @Override
+    public void seekTo(long msec) {
+        if (sMediaPlayer != null)
+            sMediaPlayer.seekTo((int)msec);
+    }
+
+    @Override
+    public void prepare(final Query query, final TomahawkMediaPlayerCallback callback) {
+        Log.d(TAG, "prepare()");
+        mMediaPlayerCallback = callback;
+        TomahawkRunnable r = new TomahawkRunnable(1) {
+            @Override
+            public void run() {
+                mPreparedQuery = null;
+                mPreparingQuery = query;
+                Result result = query.getPreferredTrackResult();
+                String path;
+                if (mTranslatedUrls.get(result) != null) {
+                    path = mTranslatedUrls.remove(result);
+                } else {
+                    if (result.getResolvedBy() instanceof ScriptResolver) {
+                        ((ScriptResolver) result.getResolvedBy()).getStreamUrl(result);
+                        return;
+                    } else {
+                        path = result.getPath();
+                    }
+                }
+                release();
+                sMediaPlayer = new MediaPlayer();
+
+                sMediaPlayer.setAudioStreamType(AudioManager.STREAM_MUSIC);
+
+                try {
+                    String finalUrl = NetworkUtils.getFinalURL(path);
+                    if (finalUrl != null)
+                        path = finalUrl;
+                } catch (IOException e) {
+                    e.printStackTrace();
```

Comment:
I think calling Log.d(TAG, "prepare - ", e); is better here. mMediaPlayerCallback.onError should also be called here